### PR TITLE
Allow overriding of auto-refresh on HTTP 403

### PR DESCRIPTION
**Pull request by bhariharan**
This PR allows the caller to override the default behavior of attempting to refresh on `HTTP 403` for a given request.

### CODE REVIEW

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java [hunk 1]
```diff
@@ -317,18 +317,18 @@ public OkHttpClient getOkHttpClient() {
      * @return
      */
     public Request buildRequest(RestRequest restRequest) {
-        Request.Builder builder =  new Request.Builder()
-                .url(HttpUrl.get(this.oAuthRefreshInterceptor.clientInfo.resolveUrl(restRequest)))
+        final Request.Builder builder =  new Request.Builder()
+                .url(HttpUrl.get(oAuthRefreshInterceptor.clientInfo.resolveUrl(restRequest)))
                 .method(restRequest.getMethod().toString(), restRequest.getRequestBody());
+        oAuthRefreshInterceptor.setShouldRefreshOn403(restRequest.getShouldRefreshOn403());
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java]: Setting whether it should refresh on `403` in the interceptor.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java [hunk 2]
```diff
@@ -338,23 +338,23 @@ public Request buildRequest(RestRequest restRequest) {
...
 	 * @return okHttp Call object (through which you can cancel the request or get the request back)
-		 */
-		public Call sendAsync(final RestRequest restRequest, final AsyncRequestCallback callback) {
-			Request request = buildRequest(restRequest);
-			Call call = okHttpClient.newCall(request);
-			call.enqueue(new Callback() {
-								 @Override
-								 public void onFailure(Call call, IOException e) {
+	 */
+    public Call sendAsync(final RestRequest restRequest, final AsyncRequestCallback callback) {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java]: Just fixed some wonky formatting here.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java [hunk 3]
```diff
@@ -618,13 +620,14 @@ public URI resolveUrl(String path) {
...
     /**
-     * Network interceptor that does oauth refresh and request retry when access token has expired
+     * Network interceptor that does oauth refresh and request retry when access token has expired.
      */
     public static class OAuthRefreshInterceptor implements Interceptor {
 
         private final AuthTokenProvider authTokenProvider;
         private String authToken;
         private ClientInfo clientInfo;
+        private boolean shouldRefreshOn403 = true;
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java]: Default is `true`.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestRequest.java [hunk 4]
```diff
@@ -162,7 +164,7 @@ public String getPath(Object... args) {
 	private final RequestBody requestBody;
 	private final Map<String, String> additionalHttpHeaders;
 	private final JSONObject requestBodyAsJson; // needed for composite and batch requests
-
+    private boolean shouldRefreshOn403 = true;
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestRequest.java]: Default is `true`.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestRequest.java [hunk 5]
```diff
@@ -320,6 +321,24 @@ public JSONObject getRequestBodyAsJson() {
...
+	public boolean getShouldRefreshOn403() {
+	    return shouldRefreshOn403;
+    }
+
+    /**
+     * Sets whether the SDK should attempt to refresh tokens if the service returns HTTP 403.
+     *
+     * @param shouldRefreshOn403 True - if the SDK should refresh on HTTP 403, False - otherwise.
+     */
+	public synchronized void setShouldRefreshOn403(boolean shouldRefreshOn403) {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestRequest.java]: The caller should call this method to override default behavior on a `RestRequest` object.

**Comment by swatikch** [comment]
I just checked quickly cherrypicking your work into the client app. Setting the flag from our end is working. More work on our end to flip the flag as needed. Thank you @bhariharan

**Pull request merged**